### ISSUE TITLE
When no refs/branches are present, tartufo doesn't actually scan anything

### ISSUE THREAD

**Issue by tarkatronic** [godaddy/tartufo#209]
## 🐛 Bug Report

The prime example of this is seen when tartufo is used by GitHub Actions, during a pull request. The checkout action will clone the target repository in a detached HEAD state, with no refs or branches present. So when tartufo goes to scan this clone, it finds no branches to scan, and simply exits, reporting success.

## To Reproduce

Set up a GitHub Action to use `actions/checkout` and then run tartufo against that clone.

## Expected Behavior

tartufo should, at a minimum, scan the current tree that is checked out. Ideally, in the case of a pull request, it should scan the entire history of that pull request, especially as it differs from the main branch.

**Comment by dgalasko-godaddy** [comment]
Bonus points if tartufo could actually error out if it can't resolve. It should not report success :)

**Comment by tarkatronic** [comment]
In the new 3.0 dev branch, with the switch to `pygit2` as a backend, this manifests differently... I recreated a clone of tartufo locally in the same form that GitHub Actions do like this:

```sh
> git init tartufo
hint: Using 'master' as the name for the initial branch. This default branch name
hint: is subject to change. To configure the initial branch name to use in all
hint: of your new repositories, which will suppress this warning, call:
hint: 
hint: 	git config --global init.defaultBranch <name>
hint: 
hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
hint: 'development'. The just-created branch can be renamed via this command:
hint: 
hint: 	git branch -m <name>
Initialized empty Git repository in /home/jwilhelm/Documents/workspace/temp/tartufo/.git/
> cd tartufo
> git config --local gc.auto 0
> git remote add origin git@github.com:godaddy/tartufo.git
> git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=1 origin +0f3fe7d5ca3a5e3d4829615e1ca7464d737285ec:refs/remotes/pull/260/merge
remote: Enumerating objects: 101, done.
remote: Counting objects: 100% (101/101), done.
remote: Compressing objects: 100% (89/89), done.
remote: Total 101 (delta 4), reused 60 (delta 1), pack-reused 0
Receiving objects: 100% (101/101), 298.03 KiB | 1.67 MiB/s, done.
Resolving deltas: 100% (4/4), done.
From github.com:godaddy/tartufo
 * [new ref]         0f3fe7d5ca3a5e3d4829615e1ca7464d737285ec -> pull/260/merge
> git checkout --progress --force refs/remotes/pull/260/merge
Note: switching to 'refs/remotes/pull/260/merge'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 0f3fe7d Merge 0aca80f8b1080c5e09524c969b4807163e5e02b2 into 9f1322988454cdda6e9d0aeb9b3c2ef358cd7851
> git branch -vvv
  * (HEAD detached at pull/260/merge) 0f3fe7d Merge 0aca80f8b1080c5e09524c969b4807163e5e02b2 into 9f1322988454cdda6e9d0aeb9b3c2ef358cd7851
```

This checkout has first helped me confirm the behavior of the old `GitPython` backend:

```sh
> tartufo -vvv scan-local-repo .
...
[2021-11-03 21:48:52,688] [INFO] - Starting scan...
[2021-11-03 21:48:53,607] [DEBUG] - Branches to be scanned: 
[2021-11-03 21:48:53,608] [INFO] - Found 0 issues.
Time: 2021-11-03T21:48:53.609486
All clear. No secrets detected.
```

No branches found, nothing scanned, no secrets detected.

Where it gets really interesting is when I start digging into this clone with `pygit2`.

```pycon
>>> from pygit2 import Repository, GIT_SORT_TOPOLOGICAL
>>> repo = Repository(".")
>>> list(repo.branches)
['pull/260/merge']
>>> branch = repo.branches.get("pull/260/merge")
>>> for commit in repo.walk(branch.resolve().target, GIT_SORT_TOPOLOGICAL):
...     print(commit)
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'object not found - no match for id (9f1322988454cdda6e9d0aeb9b3c2ef358cd7851)'
>>> 
```

This same error message is what I get when I try to run a scan with the latest v3.x:

```sh
> tartufo -vvv scan-local-repo .
[2021-11-03 22:04:47,515] [INFO] - Starting scan...
[2021-11-03 22:04:47,515] [DEBUG] - Branches to be scanned: pull/260/merge
[2021-11-03 22:04:47,515] [INFO] - Scanning branch: pull/260/merge
Traceback (most recent call last):
  File "/home/jwilhelm/Documents/workspace/tartufo/.venv/bin/tartufo", line 5, in <module>
    main()
  File "/home/jwilhelm/Documents/workspace/tartufo/.venv/lib64/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/jwilhelm/Documents/workspace/tartufo/.venv/lib64/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/jwilhelm/Documents/workspace/tartufo/.venv/lib64/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/jwilhelm/Documents/workspace/tartufo/.venv/lib64/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/jwilhelm/Documents/workspace/tartufo/.venv/lib64/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/jwilhelm/Documents/workspace/tartufo/.venv/lib64/python3.7/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/jwilhelm/Documents/workspace/tartufo/.venv/lib64/python3.7/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/jwilhelm/Documents/workspace/tartufo/tartufo/commands/scan_local_repo.py", line 50, in main
    util.process_issues(repo_path, scanner, options)
  File "/home/jwilhelm/Documents/workspace/tartufo/tartufo/util.py", line 261, in process_issues
    echo_result(options, scan, repo_path, output_dir)
  File "/home/jwilhelm/Documents/workspace/tartufo/tartufo/util.py", line 111, in echo_result
    for issue in scanner.scan():
  File "/home/jwilhelm/Documents/workspace/tartufo/tartufo/scanner.py", line 407, in scan
    for chunk in self.chunks:
  File "/home/jwilhelm/Documents/workspace/tartufo/tartufo/scanner.py", line 694, in chunks
    branch.resolve().target, pygit2.GIT_SORT_TOPOLOGICAL
KeyError: 'object not found - no match for id (9f1322988454cdda6e9d0aeb9b3c2ef358cd7851)'
```

So the difference here is that `pygit2` is seeing all refs as branches, whereas `GitPython` was not. And since `pygit2` sees a branch, but this clone is shallow (--depth=1), none of the history exists, hence the "object not found" error.

This all leads to the big question: How do we handle this problem now, with the `pygit2` backend? I'm thinking maybe I look at `repo.is_shallow` and bomb out if it's `True`, as it is in this case. Tell the user that they cannot scan the history of a shallow clone. And then if it's not a shallow clone... in theory this will just flat work, actually scanning the history of the branch, since `pygit2` sees this ref as a branch.

**Comment by tarkatronic** [comment]
Fixed in #270
